core/domain/classifier_services.py

```diff
@@ -0,0 +1,338 @@
+            probs_l[label] = probs_i[l]
+        return probs_l
+
+    def _predict(self, d, threshold=0.5):
+        """Calculates prediction data for a doc.
+        Returns a tuple of (predicted label, prediction score)."""
+        probs_l = self._get_probability_with_label(d)
+        default_prob = probs_l[self._DEFAULT_LABEL]
+        default_prob_comp = 1.0 - default_prob
+        for label in sorted(probs_l, key=probs_l.get, reverse=True):
+            prob_l = probs_l[label]
+            if (label != self._DEFAULT_LABEL and
+                    prob_l / default_prob_comp > threshold):
+                return label, prob_l / default_prob_comp
+        return self._DEFAULT_LABEL, default_prob / default_prob_comp
+
+    def _validate_label(self, label):
+        """Checks that labels don't begin with underscore."""
+        if label[0] == '_':
+            raise Exception('Label %s cannot begin with underscore.' % label)
+
+    def _parse_examples(self, examples):
+        """Splits examples into docs (split on spaces) and labels.
+        Example format:
+        [
+            ['doc1', ['label1']],
+            ['doc2', ['label2']],
+            etc
+        ]
+        """
+        docs = []
+        labels = []
+        for example in examples:
+            doc = example[0].split()
+            if len(doc) > 0:
+                docs.append(doc)
+                map(self._validate_label, example[1])
+                labels.append(example[1])
+        return docs, labels
+
+    def _train_docs(self, iterations, doc_ids=None):
+        """Trains given docs (default all) for a number of iterations."""
+        doc_ids = self._get_doc_ids(doc_ids)
+
+        for i in xrange(iterations):
+            statez = self._infer_docs(doc_ids)
+
+    def load_examples(self, examples, iterations=25):
+        """Sets new examples. Overwrites existing ones."""
+        docs, labels_list = self._parse_examples(examples)
+
+        label_set = set(
+            [self._DEFAULT_LABEL] +
+            [label for labels in labels_list for label in labels])
+
+        self._num_labels = len(label_set)
+        self._label_to_id = dict(zip(label_set, xrange(self._num_labels)))
+
+        self._num_words = 0
+        self._word_to_id = {}
+
+        self._num_docs = len(docs)
+
+        self._b_dl = numpy.array(
+            map(self._get_label_vector, labels_list), dtype=int)
+        self._w_dp = [map(self._get_word_id, doc) for doc in docs]
+        self._l_dp = []
+        self._c_dl = numpy.zeros(
+            (self._num_docs, self._num_labels), dtype=int)
+        self._c_lw = numpy.zeros(
+            (self._num_labels, self._num_words), dtype=int)
+        self._c_l = numpy.zeros(self._num_labels, dtype=int)
+
+        self._init_docs()
+        self._train_docs(iterations)
+
+    def add_examples(self, examples, iterations=25):
+        """Adds examples. Old examples are preserved.
+        Recommend 5 iterations when adding for predicting purposes.
```

Comment:
The default is set in both load and add assuming examples are training examples. The recommendation is 5 for predicting, which should only happen in add (it's not illegal to do in load, but will be extremely un-useful).

This can definitely use more doc-ing, and I think abstracting away iterations is good. Added TODO for now cause referring back to github comments isn't the fastest way to remember what I need to finish.